### PR TITLE
[Model]: Add HyperCLOVAX Audio Decoder support to vllm-omni

### PR DESCRIPTION
**Pull request by with1015**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Support audio decoder part of [HyperCLOVAX omni model](https://huggingface.co/naver-hyperclovax/HyperCLOVAX-SEED-Omni-8B) on huggingface.

(1/20) Currently, audio decoder checkpoint is available on HuggingFace in a format that is not yet supported by vllm-omni. Following internal review, it is planned to update the checkpoint to a vllm-omni–compatible format.

(4/7) Our works can be compatible with vllm-omni from current HF checkpoint.

related other PR (vision) : https://github.com/vllm-project/vllm-omni/pull/613
related issue : https://github.com/vllm-project/vllm-omni/issues/743

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by lishunyang12** [comment]
@with1015 Hey, you mentioned the HyperCLOVAX audio decoder checkpoint needs to be updated to a vllm-omni compatible format. Has that happened on the Naver side? If the checkpoint is ready, what's left on your end to get this merged?

File: vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                response.append(pcm.tobytes())
+                continue
+
+            segment = AudioSegment(pcm.tobytes(), frame_rate=24000, sample_width=pcm.dtype.itemsize, channels=1)
+
+            buf = io.BytesIO()
+            export_kwargs = {"format": fmt}
+            segment.export(buf, **export_kwargs)
+            response.append(buf.getvalue())
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py:64]: I think `post_process_func` might be missing a `return response` at the end — without it, callers would get `None` instead of the encoded audio bytes. Could you check?

File: vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+        # volume normalize
+        gain = min(volume_level / (pcm**2).mean() ** 0.5, 1 / np.max(np.abs(pcm)))
+        pcm *= gain
+        return pcm
+
+    def _load_reference_audio(self, audio: bytes, sample_rate: float) -> np.ndarray:
+        audio = io.BytesIO(audio)
+        fmt = self._detect_audio_format(audio[:4])
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py:364]: I noticed `audio` gets wrapped in `BytesIO` right above, but then `audio[:4]` tries to slice the BytesIO object — which isn't subscriptable. Maybe grab the header before wrapping?
```python
header = audio[:4]
audio = io.BytesIO(audio)
fmt = self._detect_audio_format(header)
```

File: vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+        if len(audio_tokens) != len(speakers):
+            return DiffusionOutput(output=None, error="length of speakers and audio_tokens must be the same")
+
+        # Optional: audio format. If not provided, use wav format as default.
+        formats = req.extra.get("formats", [DEFAULT_FORMAT.lower()] * len(audio_tokens))
+        if len(audio_tokens) != len(formats):
+            return DiffusionOutput(output=None, error="length of formats and audio_tokens must be the same")
+
+        ref_audio_tokens = req.extra.get("ref_audio_tokens")
+        if len(audio_tokens) != len(ref_audio_tokens):
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py:226]: If `ref_audio_tokens` is `None` (when `req.extra.get` has no default), the `len(None)` call would raise a `TypeError`. Maybe add a None guard or a default value?

File: vllm_omni/diffusion/models/hyperclovax_audio/hyperclovax_audio_decoder.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+            for layer in self.ups:
+                for l_i in layer:
+                    remove_weight_norm(l_i)
+            for layer in self.resblocks:
+                layer.remove_weight_norm()
+            remove_weight_norm(self.conv_pre)
+            remove_weight_norm(self.conv_post)
+        except ValueError:
+            print("[INFO] Model already removed weight norm. Skipping!")
+            pass
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/hyperclovax_audio_decoder.py:554]: It looks like `from_pretrained` calls `cls(h)` but the constructor expects an `OmniDiffusionConfig` — an `AttrDict` might not match. Is this method still being used, or could it be updated to match the new constructor?

File: vllm_omni/diffusion/models/hyperclovax_audio/activations.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        filter_ /= filter_.sum()
+        filter = filter_.view(1, 1, kernel_size)
+
+    return filter
+
+
+class LowPassFilter1d(nn.Module):
+    def __init__(
+        self,
+        cutoff=0.5,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/activations.py:67]: I might be misreading the flow here, but when `cutoff == 0`, it looks like the function returns `None` because `filter` only gets assigned in the `else` branch. Should the `cutoff == 0` path also return the filter? Something like:
```python
filter_ /= filter_.sum()
filter_ = filter_.view(1, 1, kernel_size)
return filter_
```

File: vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+            return None
+
+        if pad_multiple <= 0:
+            logger.warning("AUDIOLLM_PAD_MULTIPLE environment variable is not a positive int. Skipping padding...")
+            return None
+
+        return pad_multiple
+
+    def _get_pad_token_id(self) -> int | None:
+        pad_token_id_str = os.getenv("AUDIOLLM_PAD_TOKEN_ID", 3894)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py]: Just a thought — would it make sense to pull `AUDIOLLM_PAD_MULTIPLE` and `AUDIOLLM_PAD_TOKEN_ID` from the model config instead of environment variables? Env vars can be easy to forget when deploying.

File: pyproject.toml [hunk 7]
```diff
@@ -40,6 +40,7 @@ dependencies = [
     "tqdm>=4.66.0",
     "torchsde>=0.2.6",  # Required for Stable Audio scheduler
     # "vllm==0.12.0",  # TODO: fix the entrypoints overwrite problem
+    "pydub>=0.25.1",
```
Comment (lishunyang12) [inline pyproject.toml]: Since `pydub` needs `ffmpeg` as a system dependency, it might be worth adding it as an optional dependency with a helpful error message if ffmpeg isn't found. Just to make the setup experience smoother.

File: vllm_omni/diffusion/models/hyperclovax_audio/hyperclovax_audio_decoder.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+            x = sum(o for o, _ in resblock_outputs) / self.num_kernels
+            hidden_states_new.extend([h for _, h in resblock_outputs])
+
+        # Post-conv
+        x, h = self.activation_post(x, next(hidden_state_iter))
+        hidden_states_new.append(h)
+        x, h = self.conv_post(x, next(hidden_state_iter))
+        hidden_states_new.append(h)
+        # Final tanh activation
+        if self.use_tanh_at_final:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/hyperclovax_audio_decoder.py:535]: There are a few bare `print()` statements here and at lines 535, 539, 544 — the rest of the codebase uses `vllm.logger`. Might want to swap them to `logger.info` / `logger.warning` for consistency.

File: vllm_omni/diffusion/models/hyperclovax_audio/constants.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+SPEAKERS_LIST = [
+    "fkms",
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/constants.py:9]: I see `SPEAKERS_LIST` has 2 entries but `num_spk` defaults to 26 — any speaker not in the list would `KeyError`. Should the full list be included, or maybe add some validation at request time?

File: vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
+import base64
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hyperclovax_audio/pipeline_hyperclovax_audio.py:4]: Minor — a few files seem to be missing the license headers (`# SPDX-License-Identifier: Apache-2.0` etc.) that the rest of the repo has. No rush on this one.

**Review by lishunyang12 — COMMENTED** [review]
Thanks for working on this — left a few comments inline, mostly around some potential issues I spotted.

**Comment by Gaohan123** [comment]
@with1015 Hello, any updates?

**Comment by KilJaeeun** [comment]
Thank you @lishunyang12 for the thorough review! All comments have been addressed in the latest two commits (`580129cd`, `b14f193f`). Here is a summary of each fix:

**Critical bugs fixed:**
- **`post_process_func` missing `return response`** — added `return response` at the end of the inner function so callers receive encoded audio bytes instead of `None`
- **`ref_audio_tokens` None guard** — added `if ref_audio_tokens is None: ref_audio_tokens = [None] * len(audio_tokens)` before the `len()` check to avoid `TypeError`
- **`BytesIO` slicing (`audio[:4]`)** — moved header extraction before `io.BytesIO(audio)` wrapping so `_detect_audio_format` receives `bytes`, not `BytesIO`
- **`activations.py` `cutoff == 0` NameError** — moved `filter = filter_.view(1, 1, kernel_size)` outside the `else` block so it is always assigned

**Design improvements:**
- **Padding when already aligned (P2)** — added early return when `overflow == 0` to skip unnecessary full-block padding
- **Env vars → model config for `pad_multiple`/`pad_token_id`** — added these as constructor params on `HyperCLOVAXAudioDecoderModel` (stored in `self.h`); pipeline now reads from `self.bigvgan.h` instead of env vars
- **`from_pretrained` constructor mismatch** — fixed to construct `OmniDiffusionConfig` and pass `h` params as kwargs, matching the updated `__init__` signature
- **`print()` → `logger`** — all bare `print()` calls in `hyperclovax_audio_decoder.py` replaced with `logger.info` / `logger.warning`

**Minor items:**
- `SPEAKERS_LIST` vs `num_spk=26` discrepancy — added a clarifying comment in `constants.py`; the 2 exposed speaker IDs are intentional for the current release
- SPDX license headers added to `pipeline_hyperclovax_audio.py` and `constants.py`

Could you please take another look when you get a chance? @lishunyang12
